### ISSUE TITLE
Improving MMLV to VGLC conversaion

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#178]
In order for the level scene sampler to work correctly, the Mega Man levels need to have @ symbols indicating empty space, but the current script just fills that empty space with sky tiles (`-`). Add code to the script that generates VGLC ASCII levels from MMLV levels so that `@` is used in the inaccessible areas. 

For guidance, you can use the fact that level scenes have a fixed known height. I believe the actual height is 14 even though there might be code in the repo that identifies the Mega Man height as 16. If a value of 16 is in there, don't change it ... it has a purpose, but use the value 14 to help with the MMLV to VGLC conversion (include a named constant with a value of 14 ... don't use magic numbers)

**Comment by almzayena-lab** [comment]
<img width="1141" height="933" alt="Image" src="https://github.com/user-attachments/assets/03bbd34a-bef9-4105-a6ce-c6c8aa7f0973" />

<img width="1091" height="851" alt="Image" src="https://github.com/user-attachments/assets/3c3972b6-1c7a-4768-ac05-a1f8a038c09b" />

So right now i kind of fixed most of the structure to have @ in full empty space. However, now i am running into problems with the ASCII representation as it keeps some of this empty space meaning there's just too much @ being recognized now. I will try and work to find a good in between for the empty space recognition.  

**there's also that weird rectangle on that bottom not being recognized as empty.

**Comment by schrum2** [comment]
Per our conversation, the next step is to take advantage of the fact that MMLV levels are always a collection of 16 by 14 tile scenes.

**Comment by almzayena-lab** [comment]
<img width="1091" height="862" alt="Image" src="https://github.com/user-attachments/assets/a6131778-5bb4-416d-be1e-2a4f2cea7bde" />
<img width="1160" height="1079" alt="Image" src="https://github.com/user-attachments/assets/c149a7d1-9648-45b7-8feb-e4fde3722f29" />

So i got the 16 by 14 scanner to work and the text file looks amazing however now the problem is with the way the ascii is moving with it not moving fluidly and in the flow of the path it randomly shows little blocks of black the one i showed was the worst. This is not good. Im not sure what approach i will take but next class i will be working towards getting that issue resolved.

**Comment by schrum2** [comment]
Start by pinpointing exactly where within the ASCII level that particular sample came from, so that we can then investigate when it was scanned as a sample.

**Comment by almzayena-lab** [comment]
[506935.txt](https://github.com/user-attachments/files/29097892/506935.txt)

Text file i have been working on in my sampling and training process.

**Comment by schrum2** [comment]
@fwilson12 This kind of overlaps with what you are already doing. See how well your changes are able to scan the level that @almzayena-lab posted above. It is derived from an MMLV file.

**Comment by fwilson12** [comment]
https://github.com/user-attachments/assets/c6b018db-1fc5-413b-bdef-7fbca69abbc5


This is what it looks like to scroll through the level filtering for traversability (with 16x32 vertical segments and 32x16 horizontal segments).

Since the level featured two horizontal segments stacked directly on top of each other, there were a bunch of scenes during the horizontal scan that were split between the two segments, almost all of which were untraversable. 

However, it seems like levels converted from MMLV -> VGLC format can be scanned with no problems.

**Comment by schrum2** [comment]
@almzayena-lab It looks like Fletcher was able to convert that one MMLV level without issue. Please merge his branch into yours and double check if there are still issues converting that MMLV level. Once that one works, try a few others on top of that.

**Comment by schrum2** [comment]
Wanted to share a document I found that Ben made:
https://docs.google.com/document/d/1g-LgDNTVVPPGBf7P2U-3Qb_q4hlyU50VLy7X_4JNEd4/edit?usp=sharing

**Comment by schrum2** [comment]
Officially reassignin this back to @almzayena-lab

**Comment by almzayena-lab** [comment]
command were using to run

**python create_megaman_json_data. py
-- output datasets\\MM_TEST. json -- levels TEMP -- scan_mode snap -- traversable_only**

**Comment by almzayena-lab** [comment]
<img width="1093" height="944" alt="Image" src="https://github.com/user-attachments/assets/9888911d-c912-4233-a06b-9e17bef62b33" />
<img width="1092" height="941" alt="Image" src="https://github.com/user-attachments/assets/6beb0976-58b4-4607-868e-23168615e532" />

I went on to check a few other levels with this ascii and found that there were a few faulty samples. i tried to implement a filtering system that can automatically identify and keep only high-quality generated samples for dataset augmentation, but I wasn’t able to get that fully working and am trying to figure out the best approach for defining what qualifies as a “good” sample. While testing, I also realized the current A* pathfinding validation system is unreliable, since it is incorrectly marking some levels as traversable even when they should fail, so the main next steps are debugging the A* logic and better defining the sample filtering criteria for improving generation quality.

**Comment by schrum2** [comment]
The A* logic will never be perfect, but we should avoid egregiously bad cases. Don't change any of the A* logic without us discussing it first. We might want @fwilson12 to fix those problems since he implemented it, but share the examples as you see them, and I'll indicate on a case by case basis if it is worth fixing.

As for what to filter out, were the two example images above from an MMLV level? Were those particular scenes in an area of the level that is actually accessible to the player? Basically, I'm trying to figure out if the designer of the MMLV level put some level features in an inaccessible location, or if those segments only look bad because of our processing/filtering process. In particular, I wonder if these scenes resulted from sliding the window between screens. Once thing we could try is to chop the level up into 16 by 14 scenes and capture each one once with no overlap. This should work for all MMLV levels, since they have size restrictions.

To @fwilson12 : I thought that `create_megaman_json_data.py` had a setting that controlled the horizontal and vertical stride, but maybe not? If it exists, then it should solve this problem for us, but I suppose it will need to be added otherwise.

**Comment by schrum2** [comment]
@fwilson12 is working on the stride issue:
https://github.com/schrum2/MarioDiffusion/issues/180#issuecomment-4790950709
Once he is done with that, try re-scanning the levels with horizontal stride of 16 and vertical stride of 14 so that we capture all valid scenes. Note that the way we vertically pad things to a height of 16 could cause some problems, but let's see what happens after @fwilson12 commits that change

**Comment by schrum2** [comment]
Merge Fletcher's branch into yours and see if his code can process the MMLV levels properly now. See the command he applied for scanning the level one 16 by 14 scene at a time here:

https://github.com/schrum2/MarioDiffusion/issues/180#issuecomment-4791255491

**Comment by almzayena-lab** [comment]
<img width="1670" height="1079" alt="Image" src="https://github.com/user-attachments/assets/a99478e4-12f5-459d-8290-b250696aa837" />

I ran this with one level and it generated correct and accurate scenes. I will now try this with many levels (goal sample size 5,000). 
The comman i have been running it 

**python create_megaman_json_data. py -- levels TEMP -- stride_x 16 -- stride_y 14 -- scan_mode snap -- output __ stridetest. json**

**Comment by almzayena-lab** [comment]
<img width="1096" height="953" alt="Image" src="https://github.com/user-attachments/assets/a61337a0-99c0-46e9-aae0-aa09e3d0f79e" />
<img width="1085" height="946" alt="Image" src="https://github.com/user-attachments/assets/f0d81450-333e-4367-a90b-5310a0fc2439" />

found that while the dataset is generating a large number of scenes, many of the extracted samples look very poor and in some cases practically unbeatable due to awkward layouts, strange enemy placement, or sections that do not seem to reflect how the actual level should play. Because of this, my next step is to figure out a way to trace each generated scene back to the specific level and exact location it came from so I can play through those sections directly and better understand whether the issue is coming from the conversion process or the scene sampling itself.

**Comment by almzayena-lab** [comment]
[
  {
    "scene": [
      [
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1
      ],
    ....
      [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4
      ]
    ],
    "data": null,
    **"source_level": "102307.txt",**
    "source_x": 0,
    "source_y": 0,
    "scan_mode": "snap_wide"
  }

This is the update i have created in order to track where each sample comes from.

**Comment by almzayena-lab** [comment]
<img width="1206" height="1003" alt="Image" src="https://github.com/user-attachments/assets/d5ac4198-f701-485f-a81a-a6b0155a132a" />
<img width="1125" height="973" alt="Image" src="https://github.com/user-attachments/assets/6a6925f8-173f-4b85-8ec6-f4fe4753604d" />

While reviewing the generated dataset in the Tile Dataset Viewer, I noticed a number of extracted scenes looked poor or impossible to play through, often containing things like moving ground platforms (which the static tileset graphics don't represent) or an overwhelming number of enemies crammed into one screen. To investigate, I updated the code so the output JSON would tell me what level (and exact location within it) each sample came from, then manually went through and took out levels I recognized as problematic by eye. 

The pictures above the line of enemies is actually a moving ground.

**Comment by almzayena-lab** [comment]
To make this process repeatable and less manual, I added automated quality filters to create_megaman_json_data.py — --exclude_moving_ground drops any scene containing a moving-platform tile (tagged "moving" in the tileset), and --max_enemies N drops any scene with more enemy tiles than a given threshold. Both are now applied automatically during dataset generation rather than requiring manual review afterward.

use this command:
**python create_megaman_json_data.py --levels TEMP --stride_x 16 --stride_y 14 --scan_mode snap --output __stridetest.json --max_enemies 4 --min_content_pct 15**

**Comment by almzayena-lab** [comment]
After applying the updated filters (excluding moving-ground platforms by default, capping enemy density, and requiring a minimum content threshold to cut out near-empty scenes), I regenerated the dataset and ended up with around 5,000 samples. Next step is moving into training with this cleaned-up dataset.

**Comment by schrum2** [comment]
Yes, train on these with deterministic captions, and @fwilson12 can apply LLM captions to the dataset for training

**Comment by almzayena-lab** [comment]
python MM_create_ascii_captions.py --dataset __stridetest.json --tileset datasets\MM.json --output __stridetest_captioned.json --describe_absence

python tokenizer.py save --json __stridetest_captioned.json --pkl_file __stridetest_Tokenizer.pkl

python train_mlm.py --epochs 300 --save_checkpoints --json __stridetest_captioned.json --pkl __stridetest_Tokenizer.pkl --output_dir MM-MLM-stridetest --seed 0

python train_diffusion.py --pkl __stridetest_Tokenizer.pkl --json __stridetest_captioned.json --augment --mlm_model_dir MM-MLM-stridetest --text_conditional --output_dir MM_conditional_stridetest0 --seed 0 --game MM-Full

order of commands... I will also add the past commands to the readme.

**Comment by schrum2** [comment]
I got an error when running the first of these commands, presumably because I don't have the file `__stridetest.json`. Also, in general, I prefer captions without absence descriptions. This isn't as problem ... no need to cancel the training run, but in the future you should avoid using `--describe_absence`.

I would like to see more instructions in the `megaman/README.md` detailing the process of creating a Mega Man Maker diffusion model from beginning to end. The part about bulk download is there, but it says the user can use any number of levels. We should give specific recommendations based on what we are doing (and we can change the details later). In the README, walk through the process of:

1. Downloading levels
2. Converting to ASCII
3. Captioning (with deterministic captions for now, though we may change this to LLM instructions later)
    - Deterministic captions also need a tokenizer and training of MLM model 
5. Training of conditional diffusion model

Also, it seems like you are already doing this, but I just wanted to emphasize that we will always be using the MM-Full tileset moving forward.

**Comment by schrum2** [comment]
Related to the above. The file `bulk_mmlv_to_vglc.py` in the `megaman` directory does not seem to work, or I simply don't know how to use it. Instructions for this need to be in a README

**Comment by fwilson12** [comment]
Since I'm working on training models with datasets scraped from Mega Man Maker, I'll also take a look at this

**Comment by fwilson12** [comment]
Just found this .mmlv tile key, could be useful 

https://gist.github.com/freem/03d6c1832299e8d8a32f64361380920d

**Comment by fwilson12** [comment]
Recent commit refactored the mmlv->vglc script; mmlv levels now correctly can be mapped to our full tile set, and the top-left enemy bug was fixed (the top-left tile stored special data that marked the start of a new scene for scene scrolling, which the regex scanned as an enemy). 

Below are some updated samples after running bulk_download and bulk_mmlv_to_vglc that demonstrate the full tile set is now represented:

##

<img width="277" height="269" alt="Image" src="https://github.com/user-attachments/assets/10cf15d8-33a6-4a98-811f-63de4f7a2440" />

<img width="266" height="271" alt="Image" src="https://github.com/user-attachments/assets/157527f8-b897-4197-9f20-bab6a0476b19" />

<img width="267" height="274" alt="Image" src="https://github.com/user-attachments/assets/161a16de-bf70-4766-abb8-9cc3adbbe6a3" />

**Comment by almzayena-lab** [comment]
Automatic Bulk Download
Download Mega Man Maker levels by ID, The downloader begins at level ID 200000 by default (recommended to start with 100 levels).

Run:

**cd megaman
python Bulk_Download.py --target 100**

Bulk Converter (MMLV --> VGLC)
Convert downloaded .mmlv files into VGLC .txt files.

Run: 

**cd megaman
python bulk_mmlv_to_vglc.py --output ..\datasets\MM_Maker_Levels**

***This is before i check it on another device i will be checking it on another device now.

**Comment by fwilson12** [comment]
This is also related to #188, but for training on entire MMLV levels, we likely need some way to filter discontinuous levels like the one below (there are a few of these out of ~200 levels in the dataset I'm about to train on):

<img width="762" height="299" alt="Image" src="https://github.com/user-attachments/assets/ae07553c-7306-4695-b213-2edd1bd7d62d" />


## 

Additionally, there is still some weirdness with what I assume to be the moving platform block (or something similar) that is still being mapped to the default enemy tile. This is only present in 2-3 levels in the training set so I won't worry about it for now, but since I'm not using the enemy count filter for creating datasets with entire MMLV levels, the issue is present:

<img width="653" height="331" alt="Image" src="https://github.com/user-attachments/assets/560ebfe3-54c3-4af6-b228-94dfb96ed087" />

**Comment by schrum2** [comment]
There is a related issue #194 that targets some specific ways of improving this. That issue also proposes the creation of a move complete MMLV json tileset. That issue is assigned to @almzayena-lab whereas this one is assigned to both of you, but you should both be aware of waht the other is doing. 

In this issue, I would like @fwilson12 to focus on complete levels (Alaa's issue is more about individual scenes). I like the idea of tagging lots of data with metadata that will allow is to be flexible. So, when making a training set of complete levels, rather than completely eliminate discontinuous levels, let's just have an attribute in the json that labels them as discontinuous. Then, we can have a mechanism downstream that allows us to easily ignore or include the discontinuous levels.

**Comment by fwilson12** [comment]
Recent commit was supposed to reference issue #179

**Comment by schrum2** [comment]
I sort of kept changing the focus of this issue to the point where it is no longer clear what is left to do with it. I'll close this issue and clarify elsewhere.
